module/workflow/src/Domain/Entity/AbstractWorkflow.php

```diff
@@ -0,0 +1,429 @@
+use Ergonode\EventSourcing\Domain\AbstractAggregateRoot;
+use Ergonode\EventSourcing\Domain\AbstractEntity;
+use Ergonode\SharedKernel\Domain\Aggregate\TransitionId;
+use Ergonode\SharedKernel\Domain\Aggregate\WorkflowId;
+use Ergonode\Workflow\Domain\Event\Workflow\WorkflowCreatedEvent;
+use Ergonode\Workflow\Domain\Event\Workflow\WorkflowDefaultStatusSetEvent;
+use Ergonode\Workflow\Domain\Event\Workflow\WorkflowStatusAddedEvent;
+use Ergonode\Workflow\Domain\Event\Workflow\WorkflowStatusRemovedEvent;
+use Ergonode\Workflow\Domain\Event\Workflow\WorkflowTransitionAddedEvent;
+use Ergonode\Workflow\Domain\Event\Workflow\WorkflowTransitionRemovedEvent;
+use Ergonode\Workflow\Domain\ValueObject\StatusCode;
+use JMS\Serializer\Annotation as JMS;
+use Webmozart\Assert\Assert;
+use Ergonode\SharedKernel\Domain\Aggregate\StatusId;
+
+/**
+ */
+abstract class AbstractWorkflow extends AbstractAggregateRoot implements WorkflowInterface
+{
+    public const DEFAULT = 'default';
+
+    /**
+     * @var WorkflowId
+     *
+     * @JMS\Type("Ergonode\SharedKernel\Domain\Aggregate\WorkflowId")
+     */
+    private WorkflowId $id;
+
+    /**
+     * @var string
+     *
+     * @JMS\Type("string")
+     */
+    private string $code;
+
+    /**
+     * @var StatusId[]
+     *
+     * @JMS\Type("array<Ergonode\SharedKernel\Domain\Aggregate\StatusId>")
+     */
+    private array $statuses;
+
+    /**
+     * @var Transition[]
+     *
+     * @JMS\Type("array<string, Ergonode\Workflow\Domain\Entity\Transition>")
+     */
+    private array $transitions;
+
+    /**
+     * @var StatusId|null
+     *
+     * @JMS\Type("Ergonode\SharedKernel\Domain\Aggregate\StatusId")
+     */
+    private ?StatusId $defaultId;
+
+    /**
+     * @param WorkflowId $id
+     * @param string     $code
+     * @param StatusId[] $statuses
+     *
+     * @throws \Exception
+     */
+    public function __construct(WorkflowId $id, string $code, array $statuses = [])
+    {
+        Assert::allIsInstanceOf($statuses, StatusId::class);
+
+        $this->apply(new WorkflowCreatedEvent($id, get_class($this), $code, array_values($statuses)));
+    }
+
+    /**
+     * @return WorkflowId
+     */
+    public function getId(): WorkflowId
+    {
+        return $this->id;
+    }
+
+    /**
+     * @return string
+     */
+    public function getCode(): string
+    {
+        return $this->code;
+    }
+
+    /**
+     * @return string
+     */
+    abstract public function getType(): string;
+
+
+    /**
+     * @param StatusId $id
+     *
+     * @return bool
+     */
+    public function hasStatus(StatusId $id): bool
+    {
+        return isset($this->statuses[$id->getValue()]);
+    }
+
+    /**
+     * @param StatusId $id
+     *
+     * @throws \Exception
+     */
+    public function setDefaultStatus(StatusId $id): void
+    {
+        if (!$this->hasStatus($id)) {
+            throw  new \RuntimeException(sprintf('Status "%s" not exists', $id->getValue()));
```

Comment:
Domain should probably throw Domain exceptions, not global ones

Reply:
As dissused it need to be talked through on backend side